File: medicine-search-backend/src/main/java/com/medicalsystem/medicine_search/service/MedicineService.java
package com.medicalsystem.medicine_search.service;

import com.medicalsystem.medicine_search.dto.MedicineRequestDTO;
import com.medicalsystem.medicine_search.dto.MedicineSearchResponseDTO;

import java.util.List;
import java.util.Optional;

public interface MedicineService {

    // =====================
    // GET ALL MEDICINES
    // =====================
    List<MedicineSearchResponseDTO> getAllMedicines();

    // =====================
    // GET MEDICINE BY ID
    // =====================
    Optional<MedicineSearchResponseDTO> getMedicineByMedicineId (Long medicineId );

    // =============================
    // GET MEDICINES BY PRICE RANGE
    // =============================
    List<MedicineSearchResponseDTO> getMedicinesByPriceRange ( double minPrice, double maxPrice );

    // ======================
    // ✅ Save new medicine
    // ======================
    MedicineSearchResponseDTO saveMedicine ( MedicineRequestDTO medicineRequestDTO );

    // =====================
    // UPDATE MEDICINE
    // =====================
    Optional<MedicineSearchResponseDTO> updateMedicine( Long medicineId, MedicineRequestDTO medicineRequestDTO );

    // =====================
    // DELETE MEDICINE
    // =====================
    void deleteMedicine( Long medicineId );

    // ========================
    // SEARCH BY MEDICINE NAME
    // ========================
    List<MedicineSearchResponseDTO> searchMedicinesByMedicineName( String medicineName );

    // ========================================================
    // SEARCH BY NAME + PRICE RANGE (OPTIONAL - EXTRA FEATURE)
    // ========================================================
    List<MedicineSearchResponseDTO> searchMedicines( String medicineName, double minPrice, double maxPrice );
}

File: medicine-search-frontend/src/app/core/services/medicine/medicine.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, shareReplay } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../../environments/environment';
import { Medicine } from '../../../models/medicine/medicine.model';
import { MedicineRequestModel } from '../../../models/medicine-request.model/medicine-request.model';

// =====================
// API CONFIG
// =====================
const API_CONFIG = {
  BASE_URL: (window as any)['env']?.API_URL || environment.apiUrl,
};

@Injectable({
  providedIn: 'root',
})
export class MedicineService {
  // Base endpoint
  private apiUrl = `${API_CONFIG.BASE_URL}/medicine`;

  // ======
  // CACHE
  // ======
  private medicineCache$?: Observable<Medicine[]>;

  constructor(private http: HttpClient) {}

  // ===========================
  // GET ALL MEDICINES (CACHED)
  // ===========================
  getAllMedicines(): Observable<Medicine[]> {

    // 🚨 prevents repeated API calls (popup open/close will NOT re-hit backend)
    if (!this.medicineCache$) {
      this.medicineCache$ = this.http.get<Medicine[]>(this.apiUrl).pipe(
        map((res) => {
          if (Array.isArray(res)) return res;
          return [];
        }),
        shareReplay(1),
      );
    }

    return this.medicineCache$;

    // return this.http.get<Medicine[]>(this.apiUrl).pipe(
    //   map((res) => {
    //     if (Array.isArray(res)) return res;
    //     // if (res?.data) return res.data;
    //     // if (res?.medicines) return res.medicines;
    //     return [];
    //   }),
    // );
  }

  // =====================
  // GET MEDICINE BY ID
  // =====================
  getMedicineById(id: number): Observable<Medicine> {
    return this.http.get<Medicine>(`${this.apiUrl}/${id}`).pipe(
      map(res => ({
        medicineId: res.medicineId,
        medicineName: res.medicineName,
        manufacturer: res.manufacturer,
        quantity: res.quantity,
        price: res.price,
        description: res.description,
      })),
      shareReplay(1), // prevents duplicate calls if reused
    );
  }

  // =====================
  // ADD MEDICINE
  // =====================
  addMedicine(medicine: MedicineRequestModel): Observable<Medicine> {
    return this.http.post<Medicine>(this.apiUrl, medicine);
  }

  // =====================
  // UPDATE MEDICINE
  // =====================
  updateMedicine(id: number, medicine: MedicineRequestModel): Observable<Medicine> {
    return this.http.put<Medicine>(`${this.apiUrl}/${id}`, medicine);
  }

  // =====================
  // DELETE MEDICINE
  // =====================
  deleteMedicine(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // =====================
  // SEARCH MEDICINES
  // =====================
  searchMedicines( medicineName: string ): Observable<Medicine[]> {
    const url = `${this.apiUrl}/searchByName?medicineName=${encodeURIComponent( medicineName )}`;

    return this.http.get<any>(url).pipe(
      map((res) => {
        if (Array.isArray(res)) return res;
        if (res?.data) return res.data;
        if (res?.medicines) return res.medicines;
        return [];
      }),
    );
  }

  // ================================
  // REFRESH MEDICINES / RESET CACHE
  // ================================
  refreshMedicines(): Observable<Medicine[]> {
    this.medicineCache$ = undefined;
    return this.getAllMedicines();
  }

  clearCache(): void {
    this.medicineCache$ = undefined;
  }
}
